File: BinaryTreePreorderTraversal.cpp
#include <iostream>
#include <vector>
#include "Utility.h"
using namespace std;

class Solution
{
public:
    vector<int> preorderTraversal(TreeNode *root)
    {
        vector<int> result;
        vector<TreeNode*> nodeStack;
        TreeNode *p = root;
        int size;
        while (p != NULL || nodeStack.size() != 0)
        {
            if (p != NULL)
            {
                result.push_back(p->val);
                nodeStack.push_back(p);
                p = p->left;
            }
            else
            {
                size = nodeStack.size();
                p = nodeStack[size - 1];
                nodeStack.pop_back();
                p = p->right;
            }
        }
        return result;
    }
};

int main()
{
    return 0;
}

File: JavaSolution/src/BinaryTreePreorderTraversal.java
import java.util.List;
import java.util.ArrayList;

/**
 * Created by xyunpeng on 7/7/15.
 * TODO: Need to use non recursive solution
 */
public class BinaryTreePreorderTraversal {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if (root == null)
            return res;
        res.add(root.val);
        res.addAll(preorderTraversal(root.left));
        res.addAll(preorderTraversal(root.right));
        return res;
    }
}
